uwblueprint/sistering
# Built out shift signup endpoints

## Ticket link

Closes #54



## Implementation description
* Defined `CreateShiftSignupRequestDTO`, `UpdateShiftSignupRequestDTO`, `ShiftSignupResponseDTO` types
* Implemented the following CRUD endpoints with their corresponding graphql queries and mutations:
  * `createShiftSignups`
  * `updateShiftSignup`
  * `getShiftSignupsForUser` 
* Prisma schema changes (with corresponding migrations):
  * Added `note` varchar to `Signup` prisma model
  * Added foreign key constraint to `userId` in `Signup` model.
  * Added a unique key constraint to for `shiftId` and `userId` pair in `Signup` model


## Steps to test
1. Ensure that the docker containers are up and running
2. Run prisma `migration` and `generate` commands
2. Ensure that you have test data in the following tables: `users`, `shifts`, `postings`, `branches`.
  - If you don't have data, these commands might be helpful for getting some in easily:
  - Through the Postgres CLI:
    - `insert into branches (id, name) values (1, 'test branch');`
    - `insert into postings (id, branch_id, type, title, description, auto_closing_date, start_date, end_date, status) values (1, 1, 'GROUP', 'a', 'a', '2022-01-09T21:33:51.302Z', '2022-01-09T21:33:51.302Z', '2022-01-09T21:33:51.302Z', 'DRAFT');`
  - Through graphql interface (`localhost:5000/graphql`):
    - `mutation {
  createShifts(
    shifts: {
      postingId: "1"
      times: [
        { startTime: "2022-01-25T13:30", endTime: "2022-01-25T15:00" }
        { startTime: "2022-01-26T12:00", endTime: "2022-01-26T16:00" }
      ]
      endDate: "2022-03-15"
      recurrenceInterval: BIWEEKLY
    }
  ) {
    id
    postingId
    startTime
    endTime
  }
}
`
  - Credits to @lambo-liu in #55 
4. Go to `localhost:5000/graphql` and test the following new endpoints: `createShiftSignups`, `updateShiftSignup`, and `getShiftSignupsForUser`.  The following are some query/mutations that I wrote in my own testing that you might be able to use:
```
query {
  getShiftSignupsForUser(userId: "1") {
    shiftId
    userId
    numVolunteers
    note
    status
  }
}

mutation {
  createShiftSignups(
    shifts: [{ shiftId: 2, userId: 1, numVolunteers: 3, note: "test" }]
  ) {
    shiftId
    userId
    numVolunteers
    note
  }
}

mutation {
  updateShiftSignup(
    shiftId: "2"
    userId: "1"
    update: { status: CONFIRMED, note: "test", numVolunteers: 3 }
  ) {
    shiftId
    userId
    numVolunteers
    note
    status
  }
}
```

## What should reviewers focus on?
* Ensure that it works for edge cases like duplicating shift signups, signing up for a shift that doesn't exist, a user that doesn't exist signing up for a shift, etc.
* General testing for basic features


## Checklist
